File: customize.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! colored = "2.0.0"
//! str_inflector = "0.12.0"
//! dialoguer = "0.10.3"
//! handlebars = "4.3.6"
//! map-macro = "0.2.5"
//! duct = "0.13.6"
//! ```

use colored::*;
use dialoguer::{Confirm, Input};
use duct::cmd;
use handlebars::Handlebars;
use inflector::Inflector;
use map_macro::map;
use std::{env, error::Error, fs, process::ExitCode};

trait Logger {
    fn info(s: &str);
    fn error(s: &str);
    fn warning(s: &str);
}

struct Customizer {}

impl Logger for Customizer {
    fn info(s: &str) {
        println!("👉 {}", s);
    }

    fn error(s: &str) {
        eprintln!("💥 {}", s.red());
    }

    fn warning(s: &str) {
        eprintln!("🐓 {}", s.yellow());
    }
}

impl Customizer {
    fn new() -> Customizer {
        Customizer {}
    }

    fn run(&self, project_name: &str) -> Result<(), Box<dyn Error>> {
        let old_project_name_snake = "flutter_basic_quickstart";
        let old_project_name_title = "Flutter Basic Quickstart";
        let old_project_name_pascal = "FlutterBasicQuickstart";
        let project_name_snake = project_name.to_snake_case();
        let project_name_pascal = project_name.to_pascal_case();
        let project_name_title = project_name.to_title_case();
        let pubspec_yaml_path = "pubspec.yaml";
        let read_me_path = "README.md";
        let main_dart_path = "lib/main.dart";
        let test_dart_path = "test/widget_test.dart";
        let version_json5_path = "version.json5";
        let info_plist_path = "ios/Runner/Info.plist";

        fs::write(
            &pubspec_yaml_path,
            fs::read_to_string(&pubspec_yaml_path)?
                .replace(&old_project_name_snake, &project_name_snake)
                .replace(&old_project_name_title, &project_name_title),
        )?;

        fs::write(
            &main_dart_path,
            fs::read_to_string(&main_dart_path)?
                .replace(&old_project_name_pascal, &project_name_snake),
        )?;

        fs::write(
            &test_dart_path,
            fs::read_to_string(&test_dart_path)?
                .replace(&old_project_name_snake, &project_name_snake)
                .replace(&old_project_name_pascal, &project_name_pascal),
        )?;

        fs::write(
            &info_plist_path,
            fs::read_to_string(&info_plist_path)?
                .replace(&old_project_name_snake, &project_name_snake)
                .replace(&old_project_name_title, &project_name_title),
        )?;

        let description = Input::<String>::new()
            .with_prompt("Enter the description for the project")
            .interact_text()?;
        let company = Input::<String>::new()
            .with_prompt("What is the company name")
            .interact_text()?;
        let bundle_id = Input::<String>::new()
            .with_prompt("What will the iOS bundle ID be")
            .interact_text()?;

        let handlebars = Handlebars::new();

        fs::write(
            &read_me_path,
            &handlebars.render_template(
                &fs::read_to_string(&read_me_path)?,
                &map! {
                    "title" => &project_name_title,
                    "description" => &description,
                },
            )?,
        )?;

        fs::write(
            &version_json5_path,
            &handlebars.render_template(
                &fs::read_to_string(&version_json5_path)?,
                &map! {
                    "bundleId" => &bundle_id,
                    "company" => &company,
                },
            )?,
        )?;

        if Confirm::new()
            .with_prompt("Delete customization scripts?")
            .interact()?
        {
            fs::remove_file("customize.rs")?;
        }

        let reinitialize_repo = Confirm::new()
            .with_prompt("Reinitialize Git repo?")
            .interact()?;

        Self::info("Updating version information");
        cmd!("stampver", "-u", "incrBuild").run()?;

        Self::info("Downloading public packages");
        cmd!("flutter", "pub", "get").run()?;

        Self::info("Generating all build files");
        cmd!("just", "generate").run()?;

        if reinitialize_repo {
            fs::remove_dir_all(".git")?;
            cmd!("git", "init").run()?;
            cmd!("git", "add", "-A", ":/").run()?;
            cmd!("git", "commit", "-m", "'Initial commit'").run()?;
        }

        Self::info("Customization complete");

        Ok(())
    }
}

fn main() -> ExitCode {
    let args: Vec<String> = env::args().collect();
    let customizer = Customizer::new();

    if args.len() < 2 {
        eprintln!("Usage: customize [PROJECT-NAME]");
        return ExitCode::SUCCESS;
    }

    if let Err(err) = customizer.run(&args[1]) {
        Customizer::error(&err.to_string());
        return ExitCode::FAILURE;
    }

    ExitCode::SUCCESS
}

File: customize.ts
#!/usr/bin/env -S deno run --unstable --allow-run --allow-read --allow-write
import {
  Command,
  ValidationError,
} from "https://deno.land/x/cliffy@v0.25.7/command/mod.ts";
import {
  snakeCase,
  titleCase,
  pascalCase,
} from "https://deno.land/x/case@2.1.1/mod.ts";
import {
  Input,
  Confirm,
} from "https://deno.land/x/cliffy@v0.25.7/prompt/mod.ts";
import { Karacho } from "https://deno.land/x/karacho@v1.0.25/main.ts";
import * as path from "https://deno.land/std@0.184.0/path/mod.ts";

await new Command()
  .name("customize.ts")
  .version("1.0.0")
  .description("Customizes the Flutter template")
  .arguments("<PROJECTNAME:string>")
  .error((error) => {
    if (error instanceof ValidationError) {
      console.error(error.message);
      Deno.exit(1);
    }
  })
  .action(async (_options, args) => {
    const oldProjectNameSnake = "flutter_basic_quickstart";
    const oldProjectNameTitle = "Flutter Basic Quickstart";
    const oldProjectNamePascal = "FlutterBasicQuickstart";
    const oldBundleId = "com.example.flutter-basic-quickstart";
    const oldApplicationId = "com.example.flutter_basic_quickstart";
    const oldMainActivityKtPath =
      "android/app/src/main/kotlin/com/example/flutter_basic_quickstart/MainActivity.kt";
    const projectName = args;
    const projectNameSnake = snakeCase(projectName);
    const projectNamePascal = pascalCase(projectName);
    const projectNameTitle = titleCase(projectName);
    const pubspecYamlPath = "pubspec.yaml";
    const readMePath = "README.md";
    const mainDartPath = "lib/main.dart";
    const widgetTestDartPath = "test/widget_test.dart";
    const versionJson5Path = "version.json5";
    const infoPlistPath = "ios/Runner/Info.plist";
    const projectPbxproj1Path = "ios/Runner.xcodeproj/project.pbxproj";
    const buildGradlePath = "android/app/build.gradle";
    const androidManifestXml1Path = "android/app/src/debug/AndroidManifest.xml";
    const androidManifestXml2Path = "android/app/src/main/AndroidManifest.xml";
    const androidManifestXml3Path =
      "android/app/src/profile/AndroidManifest.xml";
    const cmakelistsText1Path = "linux/CMakeLists.txt";
    const myApplicationCcPath = "linux/my_application.cc";
    const appInfoXcconfigPath = "macos/Runner/Configs/AppInfo.xcconfig";
    const projectPbxproj2Path = "macos/Runner.xcodeproj/project.pbxproj";
    const runnerXcschemePath =
      "macos/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme";
    const indexHtmlPath = "web/index.html";
    const manifestJsonPath = "web/manifest.json";
    const cmakelistsText2Path = "windows/CMakeLists.txt";
    const runnerRcPath = "windows/runner/Runner.rc";
    const mainCppPath = "windows/runner/Main.cpp";
    const oldCompany = "ExampleCompany";

    const description = await Input.prompt(
      "Enter the description for the project"
    );
    const company = await Input.prompt("Enter the company name");
    let bundleId = await Input.prompt("Enter the iOS bundle ID");
    const karacho = new Karacho();

    bundleId = bundleId.replaceAll("_", "-");

    const applicationId = bundleId.replaceAll("-", "_");

    console.info(`iOS/macOS Bundle ID is '${bundleId}'`);
    console.info(`Android Application ID is '${applicationId}'`);

    // iOS
    Deno.writeTextFileSync(
      infoPlistPath,
      Deno.readTextFileSync(infoPlistPath)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
        .replaceAll(oldProjectNameTitle, projectNameTitle)
    );

    Deno.writeTextFileSync(
      projectPbxproj1Path,
      Deno.readTextFileSync(projectPbxproj1Path).replaceAll(
        oldBundleId,
        bundleId
      )
    );

    // macOS

    Deno.writeTextFileSync(
      appInfoXcconfigPath,
      Deno.readTextFileSync(appInfoXcconfigPath)
        .replaceAll(oldBundleId, bundleId)
        .replaceAll(oldCompany, company)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
    );

    Deno.writeTextFileSync(
      projectPbxproj2Path,
      Deno.readTextFileSync(projectPbxproj2Path).replaceAll(
        oldProjectNameSnake,
        projectNameSnake
      )
    );

    Deno.writeTextFileSync(
      runnerXcschemePath,
      Deno.readTextFileSync(runnerXcschemePath).replaceAll(
        oldProjectNameSnake,
        projectNameSnake
      )
    );

    // Android

    const mainActivityKtPath = `android/app/src/main/kotlin/${applicationId
      .split(".")
      .join("/")}/MainActivity.kt`;

    // Move the MainActivity.kt file into the right directory
    Deno.mkdirSync(path.dirname(mainActivityKtPath), { recursive: true });
    Deno.renameSync(oldMainActivityKtPath, mainActivityKtPath);

    if (applicationId.startsWith("com.")) {
      Deno.removeSync("android/app/src/main/kotlin/com/example", {
        recursive: true,
      });
    } else {
      Deno.removeSync("android/app/src/main/kotlin/com", { recursive: true });
    }

    Deno.writeTextFileSync(
      mainActivityKtPath,
      Deno.readTextFileSync(mainActivityKtPath).replaceAll(
        oldApplicationId,
        applicationId
      )
    );

    Deno.writeTextFileSync(
      buildGradlePath,
      Deno.readTextFileSync(buildGradlePath).replaceAll(
        oldApplicationId,
        applicationId
      )
    );

    Deno.writeTextFileSync(
      androidManifestXml1Path,
      Deno.readTextFileSync(androidManifestXml1Path).replaceAll(
        oldApplicationId,
        applicationId
      )
    );

    Deno.writeTextFileSync(
      androidManifestXml2Path,
      Deno.readTextFileSync(androidManifestXml2Path)
        .replaceAll(oldApplicationId, applicationId)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
    );

    Deno.writeTextFileSync(
      androidManifestXml3Path,
      Deno.readTextFileSync(androidManifestXml3Path).replaceAll(
        oldApplicationId,
        applicationId
      )
    );

    // Linux

    Deno.writeTextFileSync(
      cmakelistsText1Path,
      Deno.readTextFileSync(cmakelistsText1Path)
        .replaceAll(oldBundleId, bundleId)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
    );

    Deno.writeTextFileSync(
      myApplicationCcPath,
      Deno.readTextFileSync(myApplicationCcPath).replaceAll(
        oldProjectNameSnake,
        projectNameSnake
      )
    );

    // Web

    Deno.writeTextFileSync(
      indexHtmlPath,
      Deno.readTextFileSync(indexHtmlPath).replaceAll(
        oldProjectNameSnake,
        projectNameSnake
      )
    );

    Deno.writeTextFileSync(
      manifestJsonPath,
      Deno.readTextFileSync(manifestJsonPath).replaceAll(
        oldProjectNameSnake,
        projectNameSnake
      )
    );

    // Windows

    Deno.writeTextFileSync(
      cmakelistsText2Path,
      Deno.readTextFileSync(cmakelistsText2Path).replaceAll(
        oldProjectNameSnake,
        projectNameSnake
      )
    );

    Deno.writeTextFileSync(
      mainCppPath,
      Deno.readTextFileSync(mainCppPath).replaceAll(
        oldProjectNameSnake,
        projectNameSnake
      )
    );

    Deno.writeTextFileSync(
      runnerRcPath,
      Deno.readTextFileSync(runnerRcPath)
        .replaceAll(oldCompany, company)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
    );

    // General

    Deno.writeTextFileSync(
      pubspecYamlPath,
      Deno.readTextFileSync(pubspecYamlPath)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
        .replaceAll(oldProjectNameTitle, projectNameTitle)
    );

    Deno.writeTextFileSync(
      mainDartPath,
      Deno.readTextFileSync(mainDartPath)
        .replaceAll(oldProjectNameTitle, projectNameTitle)
        .replaceAll(oldProjectNamePascal, projectNamePascal)
    );

    Deno.writeTextFileSync(
      widgetTestDartPath,
      Deno.readTextFileSync(widgetTestDartPath)
        .replaceAll(oldProjectNameSnake, projectNameSnake)
        .replaceAll(oldProjectNamePascal, projectNamePascal)
    );

    Deno.writeTextFileSync(
      readMePath,
      karacho.compile(Deno.readTextFileSync(readMePath))({
        title: projectNameTitle,
        description: description,
      })
    );

    Deno.writeTextFileSync(
      versionJson5Path,
      karacho.compile(Deno.readTextFileSync(versionJson5Path))({
        company,
      })
    );

    // Clean-up

    if (await Confirm.prompt("Delete customization script?")) {
      Deno.removeSync("customize.ts");
    }

    const reinitializeRepo = await Confirm.prompt("Reinitialize Git repo?");

    // Finalization

    console.info("Copying default configuration");
    Deno.copyFileSync(
      "lib/providers/app_config_provider.default.dart",
      "lib/providers/app_config_provider.dart"
    );

    console.info("Downloading public packages");
    await new Deno.Command("flutter", { args: ["pub", "get"] }).spawn().status;

    console.info("Generating all build files");
    await new Deno.Command("just", { args: ["generate"] }).spawn().status;

    if (reinitializeRepo) {
      Deno.removeSync(".git", { recursive: true });
      await new Deno.Command("git", { args: ["init"] }).spawn().status;
      await new Deno.Command("git", { args: ["add", "-A", ":/"] }).spawn()
        .status;
      await new Deno.Command("git", {
        args: ["commit", "-m", "Initial commit"],
      }).spawn().status;
    }

    console.info("Customization complete");
  })
  .parse(Deno.args);
